File: BuiltinManager.py
from Object import *
from Utils import *
import time

class BuiltinManager(object):
    instance=None
    builtinFunctions:list[BuiltinFunctionObject]=[]
    builtinFunctionNames:list[str]=[]
    builtinVariables:list[BuiltinVariableObject]=[]
    builtinVariableNames:list[str]=[]

    def __new__(cls, *args, **kw):
        if not cls.instance:
            cls.instance = super(BuiltinManager, cls).__new__(cls, *args, **kw)
        return cls.instance
    
    def __init__(self) -> None:
        self.RegisterFunction("print",self.__Print)
        self.RegisterFunction("println",self.__Println)
        self.RegisterFunction("sizeof",self.__Sizeof)
        self.RegisterFunction("insert",self.__Insert)
        self.RegisterFunction("erase",self.__Erase)
        self.RegisterFunction("clock",self.__Clock)

    def __del__(self)->None:
        self.builtinFunctions.clear()
        self.builtinFunctionNames.clear()
        self.builtinVariables.clear()
        self.builtinVariableNames.clear()

    def RegisterFunction(self,name:str,fn:any)->None:
        self.builtinFunctions.append(BuiltinFunctionObject(name,fn))
        self.builtinFunctionNames.append(name)

    def RegisterVariable(self,name:str,obj:Object)->None:
        self.builtinVariables.append(BuiltinVariableObject(name,obj))
        self.builtinVariableNames.append(name)

    
    def __Println(self,args: list[Object]):
        if len(args) == 0:
            return False,None
        print(args[0])
        return False,None


    def __Print(self,args: list[Object]) -> bool:
        if len(args) == 0:
            return False,None
        print(args[0], end="")
        return False,None

    def __Sizeof(self,args: list[Object]):
        if len(args) == 0 or len(args) > 1:
            Assert("Native function 'siezeof':Expect a argument.")
        if args[0].type == ObjectType.ARRAY:
            result = NumObject(len(args[0].elements))
        elif args[0].type == ObjectType.STR:
            result = NumObject(len(args[0].value))
        else:
            Assert("[Native function 'sizeof']:Expect a array or string argument.")
        return True,result


    def __Insert(self,args: list[Object]):
        if len(args) == 0 or len(args) != 3:
            Assert("[Native function 'insert']:Expect 3 arguments,the arg0 must be array or string obj.The arg1 is the index obj.The arg2 is the value obj.")

        if args[0].type == ObjectType.ARRAY:
            if args[1].type != ObjectType.NUM:
                Assert(
                    "[Native function 'insert']:Arg1 must be integer type while insert to a array")
            iIndex = args[1].value
            if iIndex < 0 or iIndex >= len(args[0].elements):
                Assert("[Native function 'insert']:Index out of array's range")
            args[0].elements.insert(iIndex, args[2])
        elif args[0].type == ObjectType.STR:
            if args[1].type != ObjectType.NUM:
                Assert(
                    "[Native function 'insert']:Arg1 must be integer type while insert to a string")
            iIndex = args[1].value
            if iIndex < 0 or iIndex >= len(args[0].values):
                Assert("[Native function 'insert']:Index out of array's range")
            args[0].values.insert(iIndex, args[2].__str__())
        return False,None


    def __Erase(self,args: list[Object] ) -> bool:
        if len(args) or len(args) != 2:
            Assert("[Native function 'erase']:Expect 2 arguments,the arg0 must be array or string obj.The arg1 is the corresponding index obj.")

        if args[0].type == ObjectType.ARRAY:
            if args[1].type != ObjectType.NUM:
                Assert(
                    "[Native function 'insert']:Arg1 must be integer type while insert to a array")
            iIndex = args[1].value
            if iIndex < 0 or iIndex >= len(args[0].elements):
                Assert("[Native function 'insert']:Index out of array's range")
            args[0].elements.pop(iIndex)
        elif args[0].type == ObjectType.STR:
            if args[1].type != ObjectType.NUM:
                Assert(
                    "[Native function 'insert']:Arg1 must be integer type while insert to a string")
            iIndex = args[1].value
            if iIndex < 0 or iIndex >= len(args[0].values):
                Assert("[Native function 'insert']:Index out of array's range")
            args[0].values.pop(iIndex)
        else:
            Assert("[Native function 'erase']:Expect a array or string argument.")
        return False,None


    def __Clock(self,args: list[Object] ) -> bool:
        result = NumObject(time.perf_counter())
        return True,result

gBuiltinManager=BuiltinManager()

File: libComputeDuck/BuiltinManager.cs
using System;
using System.Collections.Generic;

namespace ComputeDuck
{
    public class BuiltinManager
    {
        private static BuiltinManager instance = null;
        public List<BuiltinFunctionObject> m_BuiltinFunctions;
        public List<string> m_BuiltinFunctionNames;
        public List<BuiltinVariableObject> m_BuiltinVariables;
        public List<string> m_BuiltinVariableNames;

        private BuiltinManager()
        {
            m_BuiltinFunctions = new List<BuiltinFunctionObject>();
            m_BuiltinFunctionNames = new List<string>();
            m_BuiltinVariables = new List<BuiltinVariableObject>();
            m_BuiltinVariableNames = new List<string>();
            Init();
        }

        public static BuiltinManager GetInstance()
        {
            if (instance == null)
                instance = new BuiltinManager();
            return instance;
        }

        public void Init()
        {
            RegisterFunction("print", _Print);
            RegisterFunction("println", _Println);
            RegisterFunction("sizeof", _SizeOf);
            RegisterFunction("insert", _Insert);
            RegisterFunction("erase", _Erase);
            RegisterFunction("clock", _Clock);
        }

        public void Release()
        {
            m_BuiltinFunctions.Clear();
            m_BuiltinFunctionNames.Clear();
            m_BuiltinVariables.Clear();
            m_BuiltinVariableNames.Clear();
        }

        public void RegisterFunction(string name, BuiltinFn fn)
        {
            m_BuiltinFunctions.Add(new BuiltinFunctionObject(name, fn));
            m_BuiltinFunctionNames.Add(name);
        }

        public void RegisterVariable(string name, Object obj)
        {
            m_BuiltinVariables.Add(new BuiltinVariableObject(name, obj));
            m_BuiltinVariableNames.Add(name);
        }

        private bool _Print(List<Object> args, out Object result)
        {
            result = null;
            if (args.Count == 0)
                return false;
            Console.Write(args[0].Stringify());
            return false;
        }

        private bool _Println(List<Object> args, out Object result)
        {
            result = null;
            if (args.Count == 0)
                return false;
            Console.WriteLine(args[0].Stringify());
            return false;
        }

        private bool _SizeOf(List<Object> args, out Object result)
        {
            result = null;
            if (args.Count == 0 || args.Count > 1)
                Utils.Assert("[Native function 'sizeof']:Expect a argument.");

            if (args[0].type == ObjectType.ARRAY)
                result = new NumObject(((ArrayObject)args[0]).elements.Count);
            else if (args[0].type == ObjectType.STR)
                result = new NumObject(((StrObject)args[0]).value.Length);
            else
                Utils.Assert("[Native function 'sizeof']:Expect a array or string argument.");
            return true;
        }

        private bool _Insert(List<Object> args, out Object result)
        {
            result = null;
            if (args.Count == 0 || args.Count != 3)
                Utils.Assert("[Native function 'insert']:Expect 3 arguments,the arg0 must be array or string object.The arg1 is the index object.The arg2 is the value object.");

            if (args[0].type == ObjectType.ARRAY)
            {
                var array = (ArrayObject)(args[0]);
                if (args[1].type != ObjectType.NUM)
                    Utils.Assert("[Native funcrion 'insert']:Arg1 must be integer type while insert to a array.");
                int iIndex = (int)((NumObject)args[1]).value;
                if (iIndex < 0 || iIndex >= array.elements.Count)
                    Utils.Assert("[Native function 'insert']:Index out of array's range");

                array.elements.Insert(iIndex, args[2]);
            }
            else if (args[0].type == ObjectType.STR)
            {
                var str = (StrObject)(args[0]);
                if (args[1].type != ObjectType.NUM)
                    Utils.Assert("[Native funcrion 'insert']:Arg1 must be integer type while insert to a string.");
                int iIndex = (int)((NumObject)args[1]).value;
                if (iIndex < 0 || iIndex >= str.value.Length)
                    Utils.Assert("[Native function 'insert']:Index out of array's range");

                str.value.Insert(iIndex, args[2].Stringify());
            }
            else
                Utils.Assert("[Native function 'insert']:Expect a array or string argument.");
            return false;
        }

        private bool _Erase(List<Object> args, out Object result)
        {
            result = null;
            if (args.Count == 0 || args.Count != 2)
                Utils.Assert("[Native function 'erase']:Expect 2 arguments,the arg0 must be array or string object..The arg1 is the corresponding index object.");

            if (args[0].type == ObjectType.ARRAY)
            {
                var array = (ArrayObject)(args[0]);
                if (args[1].type != ObjectType.NUM)
                    Utils.Assert("[Native funcrion 'erase']:Arg1 must be integer type while deleting array element.");
                int iIndex = (int)((NumObject)args[1]).value;
                if (iIndex < 0 || iIndex >= array.elements.Count)
                    Utils.Assert("[Native function 'erase']:Index out of array's range");

                array.elements.RemoveAt(iIndex);
            }
            else if (args[0].type == ObjectType.STR)
            {
                var str = (StrObject)(args[0]);
                if (args[1].type != ObjectType.NUM)
                    Utils.Assert("[Native funcrion 'erase']:Arg1 must be integer type while deleting string element.");
                 int iIndex = (int)((NumObject)args[1]).value;
                if (iIndex < 0 || iIndex >= str.value.Length)
                    Utils.Assert("[Native function 'erase']:Index out of array's range");

                str.value.Remove(iIndex);
            }
            else
                Utils.Assert("[Native function 'erase']:Expect a array or string argument.");

            return false;
        }

        private bool _Clock(List<Object> args, out Object result)
        {
            long currentTicks = DateTime.Now.Ticks;
            DateTime dtFrom = new DateTime(1970, 1, 1, 0, 0, 0, 0);
            long currentMillis = (currentTicks - dtFrom.Ticks) / 10000;
            result = new NumObject(currentMillis);
            return true;
        }
    }
}